Unhandled panic

While stress testing 500 tasks, I hit a panic

```
panic: runtime error: index out of range

goroutine 3758 [running]:
github.com/ohsu-comp-bio/funnel/util.NewDockerClient(0xc42000e468, 0x49f2366, 0x1b)
	/Users/buchanae/go/src/github.com/ohsu-comp-bio/funnel/util/docker.go:28 +0x20e
github.com/ohsu-comp-bio/funnel/worker.DockerCmd.Inspect(0xc42364b8d6, 0x6, 0xc42388d800, 0x2, 0x2, 0x4ee0d30, 0x0, 0x0, 0xc42364b924, 0xb, ...)
	/Users/buchanae/go/src/github.com/ohsu-comp-bio/funnel/worker/docker.go:92 +0xbb
github.com/ohsu-comp-bio/funnel/worker.(*stepRunner).inspectContainer(0xc4220a0300, 0x4e28740, 0xc4215f1b80)
	/Users/buchanae/go/src/github.com/ohsu-comp-bio/funnel/worker/step.go:94 +0x89
created by github.com/ohsu-comp-bio/funnel/worker.(*stepRunner).Run
	/Users/buchanae/go/src/github.com/ohsu-comp-bio/funnel/worker/step.go:50 +0x520
```

--- a/storage/gs.go
+++ b/storage/gs.go
@@ -49,9 +49,10 @@ func NewGSBackend(conf config.GSStorage) (*GSBackend, error) {
 	} else if conf.FromEnv {
 		// Pull the information (auth and other config) from the environment,
 		// which is useful when this code is running in a Google Compute instance.
-		var err error
-		client, err = google.DefaultClient(ctx, storage.CloudPlatformScope)
-		if err != nil {
+		defClient, err := google.DefaultClient(ctx, storage.CloudPlatformScope)
+		if err == nil {
+			client = defClient
+		} else {
 			log.Error("Error connecting Google Storage client. Defaulting to anonymous.", err)
 			// No auth config could be found, so default to anonymous.
 		}

--- a/storage/storage.go
+++ b/storage/storage.go
@@ -93,6 +93,7 @@ func (storage Storage) WithConfig(conf config.StorageConfig) (Storage, error) {
 	if conf.Local.Valid() {
 		local, err := NewLocalBackend(conf.Local)
 		if err != nil {
+			log.Error("Local storage backend", err)
 			return storage, err
 		}
 		storage = storage.WithBackend(local)
@@ -102,6 +103,7 @@ func (storage Storage) WithConfig(conf config.StorageConfig) (Storage, error) {
 		if c.Valid() {
 			s3, err := NewS3Backend(c)
 			if err != nil {
+				log.Error("S3 storage backend", err)
 				return storage, err
 			}
 			storage = storage.WithBackend(s3)
@@ -112,6 +114,7 @@ func (storage Storage) WithConfig(conf config.StorageConfig) (Storage, error) {
 		if c.Valid() {
 			gs, nerr := NewGSBackend(c)
 			if nerr != nil {
+				log.Error("Google Storage backend", nerr)
 				return storage, nerr
 			}
 			storage = storage.WithBackend(gs)

--- a/tests/e2e/basic_test.go
+++ b/tests/e2e/basic_test.go
@@ -7,10 +7,10 @@ import (
 )
 
 func TestHelloWorld(t *testing.T) {
-	id := run(`
+	id := fun.Run(`
     --cmd 'echo hello world'
   `)
-	task := wait(id)
+	task := fun.Wait(id)
 
 	if task.Logs[0].Logs[0].Stdout != "hello world\n" {
 		t.Fatal("Missing stdout")
@@ -21,13 +21,13 @@ func TestGetTaskView(t *testing.T) {
 	var err error
 	var task *tes.Task
 
-	id := run(`
+	id := fun.Run(`
     --cmd 'echo hello world'
     --name 'foo'
   `)
-	wait(id)
+	fun.Wait(id)
 
-	task = getView(id, tes.TaskView_MINIMAL)
+	task = fun.GetView(id, tes.TaskView_MINIMAL)
 
 	if task.Id != id {
 		t.Fatal("expected task ID in minimal view")
@@ -42,7 +42,7 @@ func TestGetTaskView(t *testing.T) {
 		t.Fatal("unexpected task logs included in minimal view")
 	}
 
-	task = getView(id, tes.TaskView_BASIC)
+	task = fun.GetView(id, tes.TaskView_BASIC)
 
 	if task.Name != "foo" {
 		t.Fatal("expected task name to be included basic view")
@@ -52,14 +52,14 @@ func TestGetTaskView(t *testing.T) {
 		t.Fatal("unexpected task logs included in basic view")
 	}
 
-	task = getView(id, tes.TaskView_FULL)
+	task = fun.GetView(id, tes.TaskView_FULL)
 
 	if task.Logs[0].Logs[0].Stdout != "hello world\n" {
 		t.Fatal("Missing stdout in full view")
 	}
 
 	// test http proxy
-	task, err = hcli.GetTask(id, "MINIMAL")
+	task, err = fun.HTTP.GetTask(id, "MINIMAL")
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -74,7 +74,7 @@ func TestGetTaskView(t *testing.T) {
 		t.Fatal("unexpected task logs included in minimal view")
 	}
 
-	task, err = hcli.GetTask(id, "BASIC")
+	task, err = fun.HTTP.GetTask(id, "BASIC")
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -87,7 +87,7 @@ func TestGetTaskView(t *testing.T) {
 		t.Fatal("unexpected task logs included in basic view")
 	}
 
-	task, err = hcli.GetTask(id, "FULL")
+	task, err = fun.HTTP.GetTask(id, "FULL")
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -106,13 +106,13 @@ func TestListTaskView(t *testing.T) {
 	var task *tes.Task
 	var err error
 
-	id := run(`
+	id := fun.Run(`
     --cmd 'echo hello world'
     --name 'foo'
   `)
-	wait(id)
+	fun.Wait(id)
 
-	tasks = listView(tes.TaskView_MINIMAL)
+	tasks = fun.ListView(tes.TaskView_MINIMAL)
 	task = tasks[0]
 
 	if task.Id == "" {
@@ -128,7 +128,7 @@ func TestListTaskView(t *testing.T) {
 		t.Fatal("unexpected task logs included in minimal view")
 	}
 
-	tasks = listView(tes.TaskView_BASIC)
+	tasks = fun.ListView(tes.TaskView_BASIC)
 	task = tasks[0]
 
 	if task.Name == "" {
@@ -139,7 +139,7 @@ func TestListTaskView(t *testing.T) {
 		t.Fatal("unexpected task logs included in basic view")
 	}
 
-	tasks = listView(tes.TaskView_FULL)
+	tasks = fun.ListView(tes.TaskView_FULL)
 	task = tasks[0]
 
 	if task.Logs[0].Logs[0].Stdout != "hello world\n" {
@@ -148,7 +148,7 @@ func TestListTaskView(t *testing.T) {
 
 	// test http proxy
 	var r *tes.ListTasksResponse
-	r, err = hcli.ListTasks("MINIMAL")
+	r, err = fun.HTTP.ListTasks("MINIMAL")
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -167,7 +167,7 @@ func TestListTaskView(t *testing.T) {
 		t.Fatal("unexpected task logs included in minimal view")
 	}
 
-	r, err = hcli.ListTasks("BASIC")
+	r, err = fun.HTTP.ListTasks("BASIC")
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -181,7 +181,7 @@ func TestListTaskView(t *testing.T) {
 		t.Fatal("unexpected task logs included in basic view")
 	}
 
-	r, err = hcli.ListTasks("FULL")
+	r, err = fun.HTTP.ListTasks("FULL")
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -196,27 +196,27 @@ func TestListTaskView(t *testing.T) {
 // This ensures that the streaming works even when a small
 // amount of logs are written (which was once a bug).
 func TestSingleCharLog(t *testing.T) {
-	id := run(`
+	id := fun.Run(`
     --cmd "sh -c 'echo a; sleep 100'"
   `)
-	waitForRunning(id)
+	fun.WaitForRunning(id)
 	time.Sleep(time.Second * 2)
-	task := get(id)
+	task := fun.Get(id)
 	if task.Logs[0].Logs[0].Stdout != "a\n" {
 		t.Fatal("Missing logs")
 	}
-	cancel(id)
+	fun.Cancel(id)
 }
 
 // Test that port mappings are being logged.
 /* TODO need ports in funnel run
 func TestPortLog(t *testing.T) {
-  id := run(`
+  id := fun.Run(`
     --cmd 'echo start'
     --cmd 'sleep 10'
   `)
-  waitForExec(id, 2)
-  task := get(id)
+  fun.WaitForExec(id, 2)
+  task := fun.Get(id)
   if task.Logs[0].Logs[0].Ports[0].Host != 5000 {
     t.Fatal("Unexpected port logs")
   }
@@ -225,15 +225,15 @@ func TestPortLog(t *testing.T) {
 
 // Test that a completed task cannot change state.
 func TestCompleteStateImmutable(t *testing.T) {
-	id := run(`
+	id := fun.Run(`
     --cmd 'echo hello'
   `)
-	wait(id)
-	err := cancel(id)
+	fun.Wait(id)
+	err := fun.Cancel(id)
 	if err == nil {
 		t.Fatal("expected error")
 	}
-	task := get(id)
+	task := fun.Get(id)
 	if task.State != tes.State_COMPLETE {
 		t.Fatal("Unexpected state")
 	}
@@ -241,17 +241,17 @@ func TestCompleteStateImmutable(t *testing.T) {
 
 // Test canceling a task
 func TestCancel(t *testing.T) {
-	id := run(`
+	id := fun.Run(`
     --cmd 'echo start'
     --cmd 'sleep 1000'
     --cmd 'echo never'
   `)
-	waitForExec(id, 1)
-	cancel(id)
+	fun.WaitForExec(id, 1)
+	fun.Cancel(id)
 	// TODO docker details and container ID are very Funnel specific
 	//      how could we generalize this to be reusued as TES conformance?
-	waitForDockerDestroy(id + "-0")
-	task := get(id)
+	fun.WaitForDockerDestroy(id + "-0")
+	task := fun.Get(id)
 	if task.State != tes.State_CANCELED {
 		t.Fatal("Unexpected state")
 	}
@@ -261,14 +261,14 @@ func TestCancel(t *testing.T) {
 // have been started or completed, i.e. steps that have yet to be started
 // won't show up in Task.Logs[0].Logs
 func TestExecutorLogLength(t *testing.T) {
-	id := run(`
+	id := fun.Run(`
     --cmd 'echo first'
     --cmd 'sleep 10'
     --cmd 'echo done'
   `)
-	waitForExec(id, 2)
-	task := get(id)
-	cancel(id)
+	fun.WaitForExec(id, 2)
+	task := fun.Get(id)
+	fun.Cancel(id)
 	if len(task.Logs[0].Logs) != 2 {
 		t.Fatal("Unexpected executor log count")
 	}
@@ -278,22 +278,22 @@ func TestExecutorLogLength(t *testing.T) {
 // the first step completed, but the correct behavior is to mark the
 // task complete after *all* steps have completed.
 func TestMarkCompleteBug(t *testing.T) {
-	id := run(`
+	id := fun.Run(`
     --cmd 'echo step 1'
     --cmd 'sleep 100'
   `)
-	waitForRunning(id)
-	waitForExec(id, 2)
-	task := get(id)
+	fun.WaitForRunning(id)
+	fun.WaitForExec(id, 2)
+	task := fun.Get(id)
 	if task.State != tes.State_RUNNING {
 		t.Fatal("Unexpected task state")
 	}
-	cancel(id)
+	fun.Cancel(id)
 }
 
 func TestTaskStartEndTimeLogs(t *testing.T) {
-	id := run(`--cmd 'echo 1'`)
-	task := wait(id)
+	id := fun.Run(`--cmd 'echo 1'`)
+	task := fun.Wait(id)
 	if task.Logs[0].StartTime == "" {
 		t.Fatal("missing task start time log")
 	}
@@ -303,9 +303,9 @@ func TestTaskStartEndTimeLogs(t *testing.T) {
 }
 
 func TestOutputFileLog(t *testing.T) {
-	dir := tempdir()
+	dir := fun.Tempdir()
 
-	id, _ := runTask(&tes.Task{
+	id, _ := fun.RunTask(&tes.Task{
 		Executors: []*tes.Executor{
 			{
 				ImageName: "alpine",
@@ -327,7 +327,7 @@ func TestOutputFileLog(t *testing.T) {
 		},
 	})
 
-	task := wait(id)
+	task := fun.Wait(id)
 	log.Debug("TEST", "task", task)
 
 	out := task.Logs[0].Outputs

--- a/tests/e2e/s3_test.go
+++ b/tests/e2e/s3_test.go
@@ -14,9 +14,9 @@ func TestMain(m *testing.M) {
 		"-p", "9999:9999",
 		"--rm",
 		"--name", "fun-minio-test",
-		"-e", "MINIO_ACCESS_KEY=" + minioKey,
-		"-e", "MINIO_SECRET_KEY=" + minioSecret,
-		"-v", storageDir + ":/export",
+		"-e", "MINIO_ACCESS_KEY=" + fun.Conf.Storage.S3[0].Key,
+		"-e", "MINIO_SECRET_KEY=" + fun.Conf.Storage.S3[0].Secret,
+		"-v", fun.StorageDir + ":/export",
 		"minio/minio", "server", "/export",
 	}
 	log.Debug("Start minio", strings.Join(args, " "))
@@ -39,11 +39,11 @@ func TestMain(m *testing.M) {
 }
 
 func TestS3(t *testing.T) {
-	id := run(`
+	id := fun.Run(`
     --cmd "sh -c 'echo foo > $out'"
     -o out=s3://bkt/test_output
   `)
-	wait(id)
+	fun.Wait(id)
 }
 
 /*

--- a/tests/e2e/storage_test.go
+++ b/tests/e2e/storage_test.go
@@ -9,13 +9,13 @@ import (
 
 // Test that a file can be passed as an input and output.
 func TestFileMount(t *testing.T) {
-	id := run(`
+	id := fun.Run(`
     --cmd "sh -c 'cat $in > $out'"
     -i in=./testdata/test_in
     -o out={{ .storage }}/test_out
   `)
-	task := wait(id)
-	c := readFile("test_out")
+	task := fun.Wait(id)
+	c := fun.ReadFile("test_out")
 	if c != "hello\n" {
 		log.Error("TASK", task)
 		log.Error("CONTENT", c)
@@ -26,16 +26,16 @@ func TestFileMount(t *testing.T) {
 // Test that the local storage system hard links input files.
 // TODO this test is unix specific because it uses syscall?
 func TestLocalFilesystemHardLinkInput(t *testing.T) {
-	writeFile("test_hard_link_input", "content")
-	id := run(`
+	fun.WriteFile("test_hard_link_input", "content")
+	id := fun.Run(`
     --cmd "echo foo"
     -i in={{ .storage }}/test_hard_link_input
   `)
-	task := wait(id)
+	task := fun.Wait(id)
 	if task.State != tes.State_COMPLETE {
 		t.Fatal("unexpected task failure")
 	}
-	name := storageDir + "/test_hard_link_input"
+	name := fun.StorageDir + "/test_hard_link_input"
 	fi, sterr := os.Lstat(name)
 	if sterr != nil {
 		panic(sterr)
@@ -51,12 +51,12 @@ func TestLocalFilesystemHardLinkInput(t *testing.T) {
 
 // Test using a symlink as an input file.
 func TestSymlinkInput(t *testing.T) {
-	id := run(`
+	id := fun.Run(`
     --cmd "sh -c 'cat $in > $out'"
     -i in=./testdata/test_in_symlink
     -o out={{ .storage }}/test_out
   `)
-	task := wait(id)
+	task := fun.Wait(id)
 	if task.State != tes.State_COMPLETE {
 		t.Fatal("Expected success on symlink input")
 	}
@@ -64,12 +64,12 @@ func TestSymlinkInput(t *testing.T) {
 
 // Test using a broken symlink as an input file.
 func TestBrokenSymlinkInput(t *testing.T) {
-	id := run(`
+	id := fun.Run(`
     --cmd "sh -c 'cat $in > $out'"
     -i in=./testdata/test_broken_symlink
     -o out={{ .storage }}/test_out
   `)
-	task := wait(id)
+	task := fun.Wait(id)
 	if task.State != tes.State_ERROR {
 		t.Fatal("Expected error on broken symlink input")
 	}
@@ -85,26 +85,26 @@ func TestBrokenSymlinkInput(t *testing.T) {
   is being tested here.
 */
 func TestSymlinkOutput(t *testing.T) {
-	id := run(`
+	id := fun.Run(`
     --cmd "sh -c 'echo foo > $dir/foo && ln -s $dir/foo $dir/sym && ln -s $dir/foo $sym'"
     -o sym={{ .storage }}/out-sym
     -O dir={{ .storage }}/out-dir
   `)
-	task := wait(id)
+	task := fun.Wait(id)
 
 	if task.State != tes.State_COMPLETE {
 		t.Fatal("expected success on symlink output")
 	}
 
-	if readFile("out-dir/foo") != "foo\n" {
+	if fun.ReadFile("out-dir/foo") != "foo\n" {
 		t.Fatal("unexpected out-dir/foo content")
 	}
 
-	if readFile("out-sym") != "foo\n" {
+	if fun.ReadFile("out-sym") != "foo\n" {
 		t.Fatal("unexpected out-sym content")
 	}
 
-	if readFile("out-dir/sym") != "foo\n" {
+	if fun.ReadFile("out-dir/sym") != "foo\n" {
 		t.Fatal("unexpected out-dir/sym content")
 	}
 }

--- a/tests/e2e/utils.go
+++ b/tests/e2e/utils.go
@@ -23,18 +23,31 @@ import (
 	"time"
 )
 
-var cli tes.TaskServiceClient
-var hcli *client.Client
 var log = logger.New("e2e")
-var rate = time.Millisecond * 1000
-var dcli *docker.Client
-var startTime = fmt.Sprintf("%d", time.Now().Unix())
-var storageDir string
-var minioKey = "AKIAIOSFODNN7EXAMPLE"
-var minioSecret = "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"
+var fun = NewFunnel()
 
 func init() {
 	logger.ForceColors()
+	fun.StartServer()
+}
+
+type Funnel struct {
+	// Clients
+	RPC    tes.TaskServiceClient
+	HTTP   *client.Client
+	Docker *docker.Client
+
+	// Config
+	Conf       config.Config
+	StorageDir string
+
+	// Internal
+	startTime string
+	rate      time.Duration
+}
+
+func NewFunnel() *Funnel {
+	var rate = time.Millisecond * 1000
 	conf := config.DefaultConfig()
 	conf = testutils.TempDirConfig(conf)
 	conf = testutils.RandomPortConfig(conf)
@@ -43,7 +56,7 @@ func init() {
 	conf.Worker.UpdateRate = rate
 	conf.ScheduleRate = rate
 
-	storageDir, _ = ioutil.TempDir("./test_tmp", "funnel-test-storage-")
+	storageDir, _ := ioutil.TempDir("./test_tmp", "funnel-test-storage-")
 	wd, _ := os.Getwd()
 
 	conf.Storage = config.StorageConfig{
@@ -53,46 +66,59 @@ func init() {
 		S3: []config.S3Storage{
 			{
 				Endpoint: "localhost:9999",
-				Key:      minioKey,
-				Secret:   minioSecret,
+				Key:      "AKIAIOSFODNN7EXAMPLE",
+				Secret:   "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY",
 			},
 		},
 	}
-
-	go server.Run(conf)
-	time.Sleep(time.Second)
 
 	conn, err := grpc.Dial(conf.RPCAddress(), grpc.WithInsecure())
 	if err != nil {
 		panic(err)
 	}
 
-	cli = tes.NewTaskServiceClient(conn)
-
 	var derr error
-	dcli, derr = util.NewDockerClient()
+	dcli, derr := util.NewDockerClient()
 	if derr != nil {
 		panic(derr)
 	}
 
-	// http client
-	hcli = client.NewClient("http://localhost:" + conf.HTTPPort)
+	return &Funnel{
+		RPC:        tes.NewTaskServiceClient(conn),
+		HTTP:       client.NewClient("http://localhost:" + conf.HTTPPort),
+		Docker:     dcli,
+		Conf:       conf,
+		StorageDir: storageDir,
+		startTime:  fmt.Sprintf("%d", time.Now().Unix()),
+		rate:       rate,
+	}
+}
+
+func (f *Funnel) Tempdir() string {
+	d, _ := ioutil.TempDir(f.StorageDir, "")
+	d, _ = filepath.Abs(d)
+	return d
+}
+
+func (f *Funnel) StartServer() {
+	go server.Run(f.Conf)
+	time.Sleep(time.Second)
 }
 
 // wait for a "destroy" event from docker for the given container ID
 // TODO probably could use docker.ContainerWait()
 // https://godoc.org/github.com/moby/moby/client#Client.ContainerWait
-func waitForDockerDestroy(id string) {
-	f := dockerFilters.NewArgs()
-	f.Add("type", "container")
-	f.Add("container", id)
-	f.Add("event", "destroy")
+func (f *Funnel) WaitForDockerDestroy(id string) {
+	fil := dockerFilters.NewArgs()
+	fil.Add("type", "container")
+	fil.Add("container", id)
+	fil.Add("event", "destroy")
 	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
 	defer cancel()
 
-	s, err := dcli.Events(ctx, dockerTypes.EventsOptions{
-		Since:   string(startTime),
-		Filters: f,
+	s, err := f.Docker.Events(ctx, dockerTypes.EventsOptions{
+		Since:   string(f.startTime),
+		Filters: fil,
 	})
 	for {
 		select {
@@ -105,15 +131,15 @@ func waitForDockerDestroy(id string) {
 }
 
 // cancel a task by ID
-func cancel(id string) error {
-	_, err := cli.CancelTask(context.Background(), &tes.CancelTaskRequest{
+func (f *Funnel) Cancel(id string) error {
+	_, err := f.RPC.CancelTask(context.Background(), &tes.CancelTaskRequest{
 		Id: id,
 	})
 	return err
 }
 
-func listView(view tes.TaskView) []*tes.Task {
-	t, err := cli.ListTasks(context.Background(), &tes.ListTasksRequest{
+func (f *Funnel) ListView(view tes.TaskView) []*tes.Task {
+	t, err := f.RPC.ListTasks(context.Background(), &tes.ListTasksRequest{
 		View: view,
 	})
 	if err != nil {
@@ -122,8 +148,8 @@ func listView(view tes.TaskView) []*tes.Task {
 	return t.Tasks
 }
 
-func getView(id string, view tes.TaskView) *tes.Task {
-	t, err := cli.GetTask(context.Background(), &tes.GetTaskRequest{
+func (f *Funnel) GetView(id string, view tes.TaskView) *tes.Task {
+	t, err := f.RPC.GetTask(context.Background(), &tes.GetTaskRequest{
 		Id:   id,
 		View: view,
 	})
@@ -134,31 +160,25 @@ func getView(id string, view tes.TaskView) *tes.Task {
 }
 
 // get a task by ID
-func get(id string) *tes.Task {
-	return getView(id, tes.TaskView_FULL)
+func (f *Funnel) Get(id string) *tes.Task {
+	return f.GetView(id, tes.TaskView_FULL)
 }
 
 // run a task and return it's ID
-func run(s string) string {
-	id, err := runE(s)
+func (f *Funnel) Run(s string) string {
+	id, err := f.RunE(s)
 	if err != nil {
 		panic(err)
 	}
 	return id
 }
 
-func tempdir() string {
-	d, _ := ioutil.TempDir(storageDir, "")
-	d, _ = filepath.Abs(d)
-	return d
-}
-
-func runE(s string) (string, error) {
+func (f *Funnel) RunE(s string) (string, error) {
 	// Process the string as a template to allow a few helpers
 	tpl := template.Must(template.New("run").Parse(s))
 	var by bytes.Buffer
 	data := map[string]string{
-		"storage": "./" + storageDir,
+		"storage": "./" + f.StorageDir,
 	}
 	if eerr := tpl.Execute(&by, data); eerr != nil {
 		return "", eerr
@@ -172,12 +192,11 @@ func runE(s string) (string, error) {
 	if len(tasks) > 1 {
 		return "", fmt.Errorf("Funnel run only handles a single task (no scatter)")
 	}
-	log.Debug("TASK", tasks[0])
-	return runTask(tasks[0])
+	return f.RunTask(tasks[0])
 }
 
-func runTask(t *tes.Task) (string, error) {
-	resp, cerr := cli.CreateTask(context.Background(), t)
+func (f *Funnel) RunTask(t *tes.Task) (string, error) {
+	resp, cerr := f.RPC.CreateTask(context.Background(), t)
 	if cerr != nil {
 		return "", cerr
 	}
@@ -185,9 +204,9 @@ func runTask(t *tes.Task) (string, error) {
 }
 
 // wait for a task to complete
-func wait(id string) *tes.Task {
-	for range time.NewTicker(rate).C {
-		t := get(id)
+func (f *Funnel) Wait(id string) *tes.Task {
+	for range time.NewTicker(f.rate).C {
+		t := f.Get(id)
 		if t.State != tes.State_QUEUED && t.State != tes.State_INITIALIZING &&
 			t.State != tes.State_RUNNING {
 			return t
@@ -197,9 +216,9 @@ func wait(id string) *tes.Task {
 }
 
 // wait for a task to be in the RUNNING state
-func waitForRunning(id string) {
-	for range time.NewTicker(rate).C {
-		t := get(id)
+func (f *Funnel) WaitForRunning(id string) {
+	for range time.NewTicker(f.rate).C {
+		t := f.Get(id)
 		if t.State == tes.State_RUNNING {
 			return
 		}
@@ -208,9 +227,9 @@ func waitForRunning(id string) {
 
 // wait for a task to reach the given executor index.
 // 1 is the first executor.
-func waitForExec(id string, i int) {
-	for range time.NewTicker(rate).C {
-		t := get(id)
+func (f *Funnel) WaitForExec(id string, i int) {
+	for range time.NewTicker(f.rate).C {
+		t := f.Get(id)
 		if len(t.Logs[0].Logs) >= i {
 			return
 		}
@@ -218,16 +237,16 @@ func waitForExec(id string, i int) {
 }
 
 // write a file to local storage
-func writeFile(name string, content string) {
-	err := ioutil.WriteFile(storageDir+"/"+name, []byte(content), os.ModePerm)
+func (f *Funnel) WriteFile(name string, content string) {
+	err := ioutil.WriteFile(f.StorageDir+"/"+name, []byte(content), os.ModePerm)
 	if err != nil {
 		panic(err)
 	}
 }
 
 // read a file from local storage
-func readFile(name string) string {
-	b, err := ioutil.ReadFile(storageDir + "/" + name)
+func (f *Funnel) ReadFile(name string) string {
+	b, err := ioutil.ReadFile(f.StorageDir + "/" + name)
 	if err != nil {
 		panic(err)
 	}

--- a/tests/e2e/utils.go
+++ b/tests/e2e/utils.go
@@ -31,6 +31,7 @@ func init() {
 	fun.StartServer()
 }
 
+// Funnel provides a test server and RPC/HTTP clients
 type Funnel struct {
 	// Clients
 	RPC    tes.TaskServiceClient
@@ -46,6 +47,8 @@ type Funnel struct {
 	rate      time.Duration
 }
 
+// NewFunnel creates a new funnel test server with some test
+// configuration automatically set: random ports, temp work dir, etc.
 func NewFunnel() *Funnel {
 	var rate = time.Millisecond * 1000
 	conf := config.DefaultConfig()
@@ -94,18 +97,22 @@ func NewFunnel() *Funnel {
 	}
 }
 
+// Tempdir returns a new temporary directory path
 func (f *Funnel) Tempdir() string {
 	d, _ := ioutil.TempDir(f.StorageDir, "")
 	d, _ = filepath.Abs(d)
 	return d
 }
 
+// StartServer starts the server
 func (f *Funnel) StartServer() {
 	go server.Run(f.Conf)
 	time.Sleep(time.Second)
 }
 
-// wait for a "destroy" event from docker for the given container ID
+// WaitForDockerDestroy waits for a "destroy" event
+// from docker for the given container ID
+//
 // TODO probably could use docker.ContainerWait()
 // https://godoc.org/github.com/moby/moby/client#Client.ContainerWait
 func (f *Funnel) WaitForDockerDestroy(id string) {
@@ -130,7 +137,7 @@ func (f *Funnel) WaitForDockerDestroy(id string) {
 	}
 }
 
-// cancel a task by ID
+// Cancel cancels a task by ID
 func (f *Funnel) Cancel(id string) error {
 	_, err := f.RPC.CancelTask(context.Background(), &tes.CancelTaskRequest{
 		Id: id,
@@ -138,6 +145,7 @@ func (f *Funnel) Cancel(id string) error {
 	return err
 }
 
+// ListView returns a task list (calls ListTasks) with the given view.
 func (f *Funnel) ListView(view tes.TaskView) []*tes.Task {
 	t, err := f.RPC.ListTasks(context.Background(), &tes.ListTasksRequest{
 		View: view,
@@ -148,6 +156,7 @@ func (f *Funnel) ListView(view tes.TaskView) []*tes.Task {
 	return t.Tasks
 }
 
+// GetView returns a task (calls GetTask) with the given view.
 func (f *Funnel) GetView(id string, view tes.TaskView) *tes.Task {
 	t, err := f.RPC.GetTask(context.Background(), &tes.GetTaskRequest{
 		Id:   id,
@@ -159,12 +168,13 @@ func (f *Funnel) GetView(id string, view tes.TaskView) *tes.Task {
 	return t
 }
 
-// get a task by ID
+// Get gets a task by ID
 func (f *Funnel) Get(id string) *tes.Task {
 	return f.GetView(id, tes.TaskView_FULL)
 }
 
-// run a task and return it's ID
+// Run uses `funnel run` syntax to create a task.
+// Run panics on error.
 func (f *Funnel) Run(s string) string {
 	id, err := f.RunE(s)
 	if err != nil {
@@ -173,6 +183,7 @@ func (f *Funnel) Run(s string) string {
 	return id
 }
 
+// RunE is like Run(), but returns an error instead of panicing.
 func (f *Funnel) RunE(s string) (string, error) {
 	// Process the string as a template to allow a few helpers
 	tpl := template.Must(template.New("run").Parse(s))
@@ -195,6 +206,7 @@ func (f *Funnel) RunE(s string) (string, error) {
 	return f.RunTask(tasks[0])
 }
 
+// RunTask calls CreateTask with the given task message and returns the ID.
 func (f *Funnel) RunTask(t *tes.Task) (string, error) {
 	resp, cerr := f.RPC.CreateTask(context.Background(), t)
 	if cerr != nil {
@@ -203,7 +215,7 @@ func (f *Funnel) RunTask(t *tes.Task) (string, error) {
 	return resp.Id, nil
 }
 
-// wait for a task to complete
+// Wait waits for a task to complete.
 func (f *Funnel) Wait(id string) *tes.Task {
 	for range time.NewTicker(f.rate).C {
 		t := f.Get(id)
@@ -215,7 +227,7 @@ func (f *Funnel) Wait(id string) *tes.Task {
 	return nil
 }
 
-// wait for a task to be in the RUNNING state
+// WaitForRunning waits for a task to be in the RUNNING state
 func (f *Funnel) WaitForRunning(id string) {
 	for range time.NewTicker(f.rate).C {
 		t := f.Get(id)
@@ -225,7 +237,7 @@ func (f *Funnel) WaitForRunning(id string) {
 	}
 }
 
-// wait for a task to reach the given executor index.
+// WaitForExec waits for a task to reach the given executor index.
 // 1 is the first executor.
 func (f *Funnel) WaitForExec(id string, i int) {
 	for range time.NewTicker(f.rate).C {
@@ -236,7 +248,7 @@ func (f *Funnel) WaitForExec(id string, i int) {
 	}
 }
 
-// write a file to local storage
+// WriteFile writes a file to the local (temporary) storage directory.
 func (f *Funnel) WriteFile(name string, content string) {
 	err := ioutil.WriteFile(f.StorageDir+"/"+name, []byte(content), os.ModePerm)
 	if err != nil {
@@ -244,7 +256,7 @@ func (f *Funnel) WriteFile(name string, content string) {
 	}
 }
 
-// read a file from local storage
+// ReadFile reads a file to the local (temporary) storage directory.
 func (f *Funnel) ReadFile(name string) string {
 	b, err := ioutil.ReadFile(f.StorageDir + "/" + name)
 	if err != nil {

--- a/tests/e2e/validation_test.go
+++ b/tests/e2e/validation_test.go
@@ -9,7 +9,7 @@ import (
 )
 
 func TestTaskNoExecutorsValidationError(t *testing.T) {
-	_, err := runTask(&tes.Task{})
+	_, err := fun.RunTask(&tes.Task{})
 	log.Debug("VALID ERR", err)
 	if err == nil {
 		t.Fatal("expected validation error")
@@ -26,7 +26,7 @@ func TestTaskNoExecutorsValidationError(t *testing.T) {
 }
 
 func TestTaskInputContentsValidationError(t *testing.T) {
-	_, err := runTask(&tes.Task{
+	_, err := fun.RunTask(&tes.Task{
 		Inputs: []*tes.TaskParameter{
 			{
 				Contents: "foo",
@@ -48,7 +48,7 @@ func TestTaskInputContentsValidationError(t *testing.T) {
 }
 
 func TestTaskInputContentsValidation(t *testing.T) {
-	_, err := runTask(&tes.Task{
+	_, err := fun.RunTask(&tes.Task{
 		Executors: []*tes.Executor{
 			{
 				ImageName: "alpine",
@@ -69,7 +69,7 @@ func TestTaskInputContentsValidation(t *testing.T) {
 }
 
 func TestTaskValidationError(t *testing.T) {
-	_, err := runTask(&tes.Task{
+	_, err := fun.RunTask(&tes.Task{
 		Executors: []*tes.Executor{
 			{},
 		},

--- a/tests/testutils/utils.go
+++ b/tests/testutils/utils.go
@@ -3,6 +3,7 @@ package testutils
 import (
 	"fmt"
 	"github.com/ohsu-comp-bio/funnel/config"
+	"github.com/ohsu-comp-bio/funnel/logger"
 	"io/ioutil"
 	"math/rand"
 	"os"
@@ -14,6 +15,7 @@ func init() {
 	// nanoseconds are important because the tests run faster than a millisecond
 	// which can cause port conflicts
 	rand.Seed(time.Now().UTC().UnixNano())
+	logger.ForceColors()
 }
 
 // RandomPort returns a random port string between 10000 and 20000.

--- a/util/docker.go
+++ b/util/docker.go
@@ -2,9 +2,11 @@ package util
 
 import (
 	"context"
+	"errors"
 	"github.com/docker/docker/client"
 	"os"
 	"regexp"
+	"time"
 )
 
 // NewDockerClient returns a new docker client. This util handles
@@ -19,10 +21,16 @@ func NewDockerClient() (*client.Client, error) {
 	// server; if not infer API version from error message and inform the client
 	// to use that version for future communication
 	if os.Getenv("DOCKER_API_VERSION") == "" {
-		_, err := dclient.ServerVersion(context.Background())
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+		defer cancel()
+
+		_, err := dclient.ServerVersion(ctx)
 		if err != nil {
 			re := regexp.MustCompile(`([0-9\.]+)`)
 			version := re.FindAllString(err.Error(), -1)
+			if version == nil {
+				return nil, errors.New("Can't connect docker client")
+			}
 			// Error message example:
 			//   Error getting metadata for container: Error response from daemon: client is newer than server (client API version: 1.26, server API version: 1.24)
 			os.Setenv("DOCKER_API_VERSION", version[1])

